From 18ea9c59e573cc6762619bd561e01e104ac12772 Mon Sep 17 00:00:00 2001
From: Andrey Albershteyn <aalbersh@kernel.org>
Date: Mon, 12 May 2025 13:26:47 +0200
Subject: [PATCH] ez: introduce --branch <BRANCH> argument for send

If you are not on the branch which you want to send b4 will fail.
This causes a bit of  inconvenience to use with tools like jj VSC as
it always works in a DETACHED HEAD state. With explicitly specified
branch, one can use b4 to send any branch.

Signed-off-by: Andrey Albershteyn <aalbersh@kernel.org>
---
 src/b4/command.py |  2 ++
 src/b4/ez.py      | 20 ++++++++++++++------
 2 files changed, 16 insertions(+), 6 deletions(-)

diff --git a/src/b4/command.py b/src/b4/command.py
index 41e7121efcf4..e93801edc0e3 100644
--- a/src/b4/command.py
+++ b/src/b4/command.py
@@ -380,6 +380,8 @@ def setup_parser() -> argparse.ArgumentParser:
     sp_send.add_argument('--cc', nargs='+', metavar='ADDR', help='Addresses to add to the Cc: list')
     sp_send.add_argument('--in-reply-to', metavar='IDENTIFIER',
                          help='Send patchset in reply to a message')
+    sp_send.add_argument('--branch', metavar='BRANCH',
+                         help='Use this this branch to work on')
     sp_send.add_argument('--not-me-too', action='store_true', default=False,
                          help='Remove yourself from the To: or Cc: list')
     sp_send.add_argument('--resend', metavar='vN', nargs='?', const='latest',
diff --git a/src/b4/ez.py b/src/b4/ez.py
index ab7b7eab913d..6c652b65304e 100644
--- a/src/b4/ez.py
+++ b/src/b4/ez.py
@@ -741,6 +741,8 @@ def find_cover_commit(usebranch: Optional[str] = None) -> Optional[str]:
     if usebranch is None:
         logger.critical("The current repository is not tracking a branch. To use b4, please checkout a branch.")
         logger.critical("Maybe a rebase is running?")
+        import traceback
+        traceback.print_stack()
         raise RuntimeError("Not currently on a branch, please checkout a b4-tracked branch")
 
     # Restrict to committer being the current person, in case an errant cover letter
@@ -1752,13 +1754,18 @@ def cmd_send(cmdargs: argparse.Namespace) -> None:
         auth_verify(cmdargs)
         return
 
-    mybranch = b4.git_get_current_branch()
+    if cmdargs.branch:
+        mybranch = cmdargs.branch
+    else:
+        mybranch = b4.git_get_current_branch()
+
+    print(f"branch: {mybranch}")
 
     config = b4.get_main_config()
 
     tag_msg = None
     cl_msgid = None
-    cover, tracking = load_cover(strip_comments=True)
+    cover, tracking = load_cover(strip_comments=True, usebranch=mybranch)
     if cmdargs.resend:
         if cmdargs.resend == 'latest':
             revstr = tracking['series']['revision'] - 1
@@ -1797,7 +1804,8 @@ def cmd_send(cmdargs: argparse.Namespace) -> None:
             prefixes = None
 
         try:
-            todests, ccdests, tag_msg, patches = get_prep_branch_as_patches(prefixes=prefixes)
+            todests, ccdests, tag_msg, patches = get_prep_branch_as_patches(
+                    usebranch=mybranch, prefixes=prefixes)
         except RuntimeError as ex:
             logger.critical('CRITICAL: Failed to convert range to patches: %s', ex)
             sys.exit(1)
@@ -2166,7 +2174,7 @@ def reroll(mybranch: str, tag_msg: str, msgid: str, tagprefix: str = SENT_TAG_PR
     if len(chunks) > 1:
         tag_msg = chunks[0] + '\n'
 
-    cover, tracking = load_cover(strip_comments=True)
+    cover, tracking = load_cover(strip_comments=True, usebranch=mybranch)
     revision = tracking['series']['revision']
     change_id = tracking['series']['change-id']
 
@@ -2215,7 +2223,7 @@ def reroll(mybranch: str, tag_msg: str, msgid: str, tagprefix: str = SENT_TAG_PR
                         logger.critical(out.strip())
                         raise RuntimeError
             elif strategy == 'tip-commit':
-                cover_commit = find_cover_commit()
+                cover_commit = find_cover_commit(usebranch=mybranch)
                 tagcommit = f'{cover_commit}~1'
 
             logger.info('Tagging %s', tagname)
@@ -2233,7 +2241,7 @@ def reroll(mybranch: str, tag_msg: str, msgid: str, tagprefix: str = SENT_TAG_PR
         logger.info('NOTE: Tagname %s already exists', tagname)
 
     logger.info('Recording series message-id in cover letter tracking')
-    cover, tracking = load_cover(strip_comments=False)
+    cover, tracking = load_cover(strip_comments=False, usebranch=mybranch)
     vrev = f'v{revision}'
     if 'history' not in tracking['series']:
         tracking['series']['history'] = dict()
-- 
2.47.2

